Mathlib/Algebra/Algebra/Subalgebra/Unitization.lean

```diff
@@ -12,86 +12,162 @@ import Mathlib.Algebra.Star.NonUnitalSubalgebra
+    (s : S) (h : ∀ r, r ≠ 0 → algebraMap R A r ∉ s) [AlgHomClass F R (Unitization R s) A] (f : F)
+    (hf : ∀ x : s, f x = x) : Function.Injective f := by
   refine' (injective_iff_map_eq_zero _).mpr fun x hx => _
   induction' x using Unitization.ind with r a
-  rw [map_add] at hx
-  have hx' := congr_arg (Subtype.val : _ → A) hx
-  simp only [NonUnitalSubalgebra.unitization_apply_coe, Unitization.fst_inl,
-    Subalgebra.coe_algebraMap, Unitization.snd_inl, ZeroMemClass.coe_zero, add_zero, map_neg,
-    AddSubgroupClass.coe_neg, Unitization.fst_inr, map_zero, Unitization.snd_inr,
-    Subalgebra.coe_add, zero_add] at hx'
+  simp_rw [map_add, hf, ←Unitization.algebraMap_eq_inl, AlgHomClass.commutes] at hx
+  rw [add_eq_zero_iff_eq_neg] at hx ⊢
   by_cases hr : r = 0
-  · simp only [hr, map_zero, Unitization.inl_zero, zero_add] at hx' ⊢
-    rw [← ZeroMemClass.coe_zero s] at hx'
-    exact congr_arg _ (Subtype.coe_injective hx')
-  · refine' False.elim (h1 _)
-    rw [← eq_sub_iff_add_eq, zero_sub] at hx'
-    replace hx' := congr_arg (fun y => r⁻¹ • y) hx'
-    simp only [Algebra.algebraMap_eq_smul_one, ← smul_assoc, smul_eq_mul, inv_mul_cancel hr,
-      one_smul] at hx'
-    exact hx'.symm ▸ SMulMemClass.smul_mem r⁻¹ (neg_mem a.prop)
+  · ext <;> simp [hr] at hx ⊢
+    exact hx
+  · exact (h r hr <| hx ▸ (neg_mem a.property)).elim
+
+/-- This is a generic version which allows us to prove both
+`NonUnitalSubalgebra.unitization_injective` and `NonUnitalStarSubalgebra.unitization_injective`. -/
+theorem AlgHomClass.unitization_injective {F R S A : Type _} [Field R] [Ring A]
+    [Algebra R A] [SetLike S A] [hSA : NonUnitalSubringClass S A] [hSRA : SMulMemClass S R A]
+    (s : S) (h1 : 1 ∉ s) [AlgHomClass F R (Unitization R s) A] (f : F)
+    (hf : ∀ x : s, f x = x) : Function.Injective f := by
+  refine AlgHomClass.unitization_injective' s (fun r hr hr' ↦ ?_) f hf
+  rw [Algebra.algebraMap_eq_smul_one] at hr'
+  exact h1 <| inv_smul_smul₀ hr (1 : A) ▸ SMulMemClass.smul_mem r⁻¹ hr'
+
+variable {R S A : Type _} [Field R] [Ring A] [Algebra R A]
+  [SetLike S A] [hSA : NonUnitalSubringClass S A] [hSRA : SMulMemClass S R A] (s : S)
+
+theorem NonUnitalSubalgebra.unitization_injective (h1 : (1 : A) ∉ s) :
+    Function.Injective (NonUnitalSubalgebra.unitization s) :=
+  AlgHomClass.unitization_injective s h1 (NonUnitalSubalgebra.unitization s) fun _ ↦ by simp
 
 /-- If a `NonUnitalSubalgebra` over a field does not contain `1`, then its unitization is
 isomorphic to its `Algebra.adjoin`. -/
-noncomputable def NonUnitalSubalgebra.unitizationAlgEquiv {R S A : Type _} [Field R] [Ring A]
-    [Algebra R A] [SetLike S A] [NonUnitalSubringClass S A] [SMulMemClass S R A]
-    (s : S) (h1 : (1 : A) ∉ s) : Unitization R s ≃ₐ[R] Algebra.adjoin R (s : Set A) :=
-  AlgEquiv.ofBijective (NonUnitalSubalgebra.unitization s)
-    ⟨NonUnitalSubalgebra.unitization_injective s h1, NonUnitalSubalgebra.unitization_surjective s⟩
+@[simps!]
```

Comment:
```suggestion
@[simps! apply_coe]
```
The generated `symm_apply` lemma is at best completely useless, at worst simp-set poisoning